clusterloader2/pkg/measurement/common/simple/wait_for_services.go

```diff
@@ -0,0 +1,147 @@
+	return &waitForServicesMeasurement{}
+}
+
+type waitForServicesMeasurement struct {
+	clientSet     clientset.Interface
+	namespace     string
+	labelSelector string
+	fieldSelector string
+	waitTimeout   time.Duration
+	store         *measurementutil.ServiceStore
+}
+
+// Execute waits until services are ready.
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all-namespace scope is assumed.
+// "Start" action starts observation of the services, while "gather" waits until
+// desired number of services are ready.
+func (w *waitForServicesMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		w.clientSet = config.ClusterFramework.GetClientSets().GetClient()
+		w.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return nil, err
+		}
+		w.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		w.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		w.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultWaitForServicesTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, w.start()
+	case "gather":
+		desiredCount, err := util.GetInt(config.Params, "desiredCount")
+		if err != nil {
+			return nil, err
+		}
+		return nil, w.gather(desiredCount)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (w *waitForServicesMeasurement) start() error {
+	var err error
+	w.store, err = measurementutil.NewServiceStore(w.clientSet, w.namespace, w.labelSelector, w.fieldSelector)
+	if err != nil {
+		return fmt.Errorf("service store creation error: %v", err)
+	}
+	return nil
+}
+
+func (w *waitForServicesMeasurement) gather(desiredCount int) error {
+	if w.store == nil {
+		klog.Errorf("%s: measurement is not started", w)
+		return fmt.Errorf("no service store")
+	}
+
+	created, available := 0, 0
+	selectorsString := measurementutil.CreateSelectorsString(w.namespace, w.labelSelector, w.fieldSelector)
+	for {
+		select {
+		case <-time.After(w.waitTimeout):
+			return fmt.Errorf("timeout while waiting for %d services to be available in namespace '%v' with labels '%v' and fields '%v' - only %d found running",
+				desiredCount, w.namespace, w.labelSelector, w.fieldSelector, available)
+		case <-time.After(defaultWaitForServicesInterval):
+			services := w.store.List()
+			created = len(services)
+			available = 0
+			// TODO: Measurement waits only for load balancers.
+			// It should be extended to all kinds of services.
+			for i := 0; i < len(services); i++ {
+				if len(services[i].Status.LoadBalancer.Ingress) > 0 {
```

Comment:
@prameshj @bowei - one thing that I'm not sure about and would need your clarification is:
- does the fact that LoadBalancerStatus has the Ingress set already means that the LB is programmed?
- or maybe (which is what I thought) it means that we chose the ip address, and we started programming it

Reply:
Since the IP has been populated in status, the loadbalancer has been setup at this point. I have seen there being some delay after this in actually getting a response when pinging the LB ip. Hence the need to measure this latency.